pkg/volume/util/operationexecutor/node_expander.go

```diff
@@ -79,6 +83,15 @@ func (ne *NodeExpander) runPreCheck() bool {
 		ne.resizeStatus = currentStatus
 	}
 
+	pvcSpecCap := ne.pvc.Spec.Resources.Requests[v1.ResourceStorage]
+	// usually when are performing node expansion, we expect pv size and pvc spec size
+	// to be the same, but if user has edited pvc since then and volume expansion failed
+	// with final error, then we should let controller reconcile this state.
+	if pvcSpecCap.Cmp(ne.pluginResizeOpts.NewSize) != 0 &&
+		ne.actualStateOfWorld.CheckVolumeInFailedExpansionWithFinalErrors(ne.vmt.VolumeName) {
+		ne.markExpansionInfeasible = true
+	}
```

Comment:
The comment does not help me to understand the code.

I _think_ the code makes sure that node stops retrying, so the controller can reconcile to the new size.

But why the check cannot be done after `volumePlugin.NodeExpand` in `expandOnPlugin()`? Will `expandOnPlugin()`, with a final error, modify the PV/PVC so the check cannot be used there?

Reply:
I have updated the comment and renamed the variable. I am mainly keeping this in `NodeExpander` - to keep all the state related to this resize operation self-contained. Code in operation_generator is somewhat hard to test, but `node_expander` isn't.